src/globals.c

```diff
@@ -1080,214 +1063,260 @@ void print_indent(int indent)
+                   rs2->var_name);
             break;
         case OP_log_or:
-            print_indent(indent);
-            printf("%%%s = or %%%s, %%%s", rd, op1, op2);
+            print_indent(1);
+            printf("%%%s = or %%%s, %%%s", rd->var_name, rs1->var_name,
+                   rs2->var_name);
             break;
         case OP_log_not:
-            print_indent(indent);
-            printf("%%%s = not %%%s", rd, op1);
+            print_indent(1);
+            printf("%%%s = not %%%s", rd->var_name, rs1->var_name);
             break;
         case OP_rshift:
-            print_indent(indent);
-            printf("%%%s = rshift %%%s, %%%s", rd, op1, op2);
+            print_indent(1);
+            printf("%%%s = rshift %%%s, %%%s", rd->var_name, rs1->var_name,
+                   rs2->var_name);
             break;
         case OP_lshift:
-            print_indent(indent);
-            printf("%%%s = lshift %%%s, %%%s", rd, op1, op2);
+            print_indent(1);
+            printf("%%%s = lshift %%%s, %%%s", rd->var_name, rs1->var_name,
+                   rs2->var_name);
             break;
         default:
+            printf("<Unsupported opcode: %d>", insn->opcode);
             break;
         }
+
         printf("\n");
     }
-    printf("===\n");
+}
+
+void dump_bb_insn_by_dom(func_t *func, basic_block_t *bb, bool *at_func_start)
+{
+    dump_bb_insn(func, bb, at_func_start);
+    for (int i = 0; i < MAX_BB_DOM_SUCC; i++) {
+        if (!bb->dom_next[i])
+            break;
+        dump_bb_insn_by_dom(func, bb->dom_next[i], at_func_start);
+    }
+}
+
+void dump_insn()
+{
+    for (func_t *func = FUNC_LIST.head; func; func = func->next) {
+        bool at_func_start = true;
+
+        printf("def %s", func->return_def.type_name);
+
+        for (int i = 0; i < func->return_def.is_ptr; i++)
+            printf("*");
+        printf(" @%s(", func->return_def.var_name);
+
+        for (int i = 0; i < func->num_params; i++) {
+            if (i != 0)
+                printf(", ");
+            printf("%s", func->param_defs[i].type_name);
+
+            for (int k = 0; k < func->param_defs[i].is_ptr; k++)
+                printf("*");
+            printf(" %%%s", func->param_defs[i].var_name);
+        }
+        printf(") {\n");
+
+        dump_bb_insn_by_dom(func, func->bbs, &at_func_start);
+
+        /* Handle implicit return */
+        for (int i = 0; i < MAX_BB_PRED; i++) {
+            basic_block_t *bb = func->exit->prev[i].bb;
+            if (!bb)
+                continue;
+
+            if (strcmp(func->return_def.type_name, "void"))
+                continue;
+
+            if (bb->insn_list.tail)
+                if (bb->insn_list.tail->opcode == OP_return)
+                    continue;
+
+            print_indent(1);
+            printf("ret\n");
+        }
+
+        printf("}\n");
+    }
+
+    printf("==<END OF INSN DUMP>==\n");
```

Comment:
Where is the beginning?

Can you follow the way of Linux? See https://www.kernel.org/doc/html/v5.1/admin-guide/bug-hunting.html

Reply:
Should we also do the same thing in phase 2 IR?

Reply:
> Should we also do the same thing in phase 2 IR?

Not necessary. We shall concentrate on the removal of phase 1 IR.